src/main.cpp

```diff
@@ -476,16 +479,16 @@ void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) {
 }
 
 void MaybeSetPeerAsAnnouncingHeaderAndIDs(const CNodeState* nodestate, CNode* pfrom) {
-    if (nLocalServices & NODE_WITNESS) {
-        // Don't ever request compact blocks when segwit is enabled.
+    if (nLocalServices & ~pfrom->nServices & NODE_WITNESS) {
+        // Never ask from peers who can't provide witnesses.
```

Comment:
This means that prior to segwit activation, blocks found by unupgraded miners will propagate without the benefit of compactblock announcements to upgraded nodes.

Perhaps it'd be better to instead use the BIP9 LOCKED_IN window to transition from receiving compactblock announcements from any peer regardless of nServices, towards NODE_WITNESS peers, and once ACTIVE state is reached, we enforce that all announcing peers are NODE_WITNESS?

To be concrete, here's one suggestion:
 1. before LOCKED_IN, use existing algorithm
 1. after LOCKED_IN, don't allow NODE_WITNESS peers to be evicted by non-WITNESS peers, and preferentially first evict non-WITNESS peers when trying to add a WITNESS peer
 1. after ACTIVATED, evict any non-WITNESS peers we have left making announcements.

I think this ought to provide a smooth transition, but I actually am not sure we even need that second step, we can probably be a little more disruptive around the LOCKED_IN time to ensure that we're ready at activation.